### PR TITLE
feat: move `getEntities` ordering to database

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

instead sorting the entities in the client on demand, use default ordering parameters for entities when calling `getEntities`. having thousands of entities causes this function to pop up as the sorting method is not very lightweight thus moving it to database to handle.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by drodil** [comment]
We have around 25k items in the catalog. Especially when fetching without any filters, the `refCompare` pops up in the profiling. 

Here's a short profile (~75000ms) running our instance with at least some calls to `getEntities` without ordering:
image

After adding `order` field to all of our `getEntities` calls the `refCompare` is obviously gone from the profile and more time is spent on idle. Probably this requires also a comparison from the database POV (will get back to that later). 

There were some performance tests in the backstage repository, maybe those could be utilized also to test this?

**Comment by drodil** [comment]
So, I made a performance test (see here: https://gist.github.com/drodil/c23109aea28850a8cfef39ac0ff267f0), and here are the results:

**8000 entities (master)**
- [POSTGRES_16] Fetched 8000 entities in 1041ms
- [SQLITE_3] Fetched 8000 entities in 988ms

**28000 entities (master)**
- [POSTGRES_16] Fetched 28000 entities in 4109ms
- [SQLITE_3] Fetched 28000 entities in 4228ms

**8000 entities (get_entities_db_order)**
- [POSTGRES_16] Fetched 8000 entities in 209ms
- [SQLITE_3] Fetched 8000 entities in 38577ms

**28000 entities (get_entities_db_order)**
- [POSTGRES_16] Fetched 28000 entities in 654ms
- [SQLITE_3]: Timed out after 600000ms

In conclusion, this patch slows down SQLite use case **a lot** but Postgres works multiple times faster. Nobody runs Backstage on top of SQLite outside the development environment so I don't worry about that too much. Of course, there might be something to speed this up for SQLite, for example, making the sorting in code instead of SQL.

WDYT? 

On a side note: I did not include the performance test in this PR since it was done using the existing `catalog-backend` performance test features while it's actually testing the catalog client.

**Comment by drodil** [comment]
Another option would be to completely remove the default ordering introduced in this PR and let the entities come in any order. I don't know if this breaks some features though...

**Comment by drodil** [comment]
Something like this:

```diff
diff --git a/packages/catalog-client/src/CatalogClient.ts b/packages/catalog-client/src/CatalogClient.ts
index 171b8b07527..1f611b05c88 100644
--- a/packages/catalog-client/src/CatalogClient.ts
+++ b/packages/catalog-client/src/CatalogClient.ts
@@ -126,12 +126,6 @@ export class CatalogClient implements CatalogApi {
       }
     }
 
-    if (encodedOrder.length === 0) {
-      encodedOrder.push(`asc:kind`);
-      encodedOrder.push(`asc:metadata.namespace`);
-      encodedOrder.push(`asc:metadata.name`);
-    }
-
     const entities = await this.requestRequired(
       await this.apiClient.getEntities(
         {
diff --git a/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts b/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts
index 53e8d4c1cd3..da2b82612f9 100644
--- a/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts
+++ b/plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.ts
@@ -14,7 +14,11 @@
  * limitations under the License.
  */
 
-import { DEFAULT_NAMESPACE, Entity } from '@backstage/catalog-model';
+import {
+  DEFAULT_NAMESPACE,
+  Entity,
+  stringifyEntityRef,
+} from '@backstage/catalog-model';
 import { InputError } from '@backstage/errors';
 import { DbSearchRow } from '../../tables';
 
@@ -182,6 +186,7 @@ export function buildEntitySearch(
 
   // Start with some special keys that are always present because you want to
   // be able to easily search for null specifically
+  raw.push({ key: 'entityRef', value: stringifyEntityRef(entity) });
   raw.push({ key: 'metadata.name', value: entity.metadata.name });
   raw.push({ key: 'metadata.namespace', value: entity.metadata.namespace });
   raw.push({ key: 'metadata.uid', value: entity.metadata.uid });
diff --git a/plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts b/plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts
index 9049685d389..22d9470b691 100644
--- a/plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts
+++ b/plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts
@@ -209,7 +209,11 @@ export class DefaultEntitiesCatalog implements EntitiesCatalog {
     let entitiesQuery =
       db<DbFinalEntitiesRow>('final_entities').select('final_entities.*');
 
-    request?.order?.forEach(({ field }, index) => {
+    const ordering: EntityOrder[] = request?.order ?? [
+      { field: 'entityRef', order: 'asc' },
+    ];
+
+    ordering.forEach(({ field }, index) => {
       const alias = `order_${index}`;
       entitiesQuery = entitiesQuery.leftOuterJoin(
         { [alias]: 'search' },
@@ -233,7 +237,7 @@ export class DefaultEntitiesCatalog implements EntitiesCatalog {
       );
     }
 
-    request?.order?.forEach(({ order }, index) => {
+    ordering.forEach(({ order }, index) => {
       if (db.client.config.client === 'pg') {
         // pg correctly orders by the column value and handling nulls in one go
         entitiesQuery = entitiesQuery.orderBy([
@@ -249,6 +253,7 @@ export class DefaultEntitiesCatalog implements EntitiesCatalog {
         ]);
       }
     });
+
     entitiesQuery = entitiesQuery.orderBy('final_entities.entity_id', 'asc'); // stable sort
 
     const { limit, offset } = parsePagination(request?.pagination);
```

**Comment by drodil** [comment]
Latest results with using the `refresh_state.entity_ref` for stable sort:

- [POSTGRES_16] Fetched 8000 entities in 232ms
- [SQLITE_3] Fetched 8000 entities in 109ms

- [POSTGRES_16] Fetched 28000 entities in 531ms
- [SQLITE_3] Fetched 28000 entities in 420ms

**Review by Rugvip — APPROVED** [review]
Nice 😎 :shipit:

**Review by freben — APPROVED** [review]
